### PR TITLE
Make email templates translatable

### LINKED ISSUE #5263: Make email templates translatable
Email templates are the one last thing in Indico that's basically untranslated; the templates have no i18n markers and are only in english, but some users require them to be translated as well, and as e.g. #5062 indicates this is a major hassle right now.

Main tasks here:

- Add translation markers to email templates; be especially careful in case of plaintext emails to not break the formatting. Ideally, a unit test with a snapshot is added for each of these email templates (see `test_move_request_creation_email_plaintext` for an example) before touching them to ensure the formatting/spacing does not change.
- Provide some way (e.g. a contextmanager) to temporarily switch the active language, either to a specific language or to that of a given user.
- Use this whenever we send emails to use the recipient's language if there is one or (in case the recipient is not an Indico user) a reasonable default (`DEFAULT_LOCALE`)

AFAIK registration emails currently interpolate a title-cased enum key name somewhere, iirc. for the state. This probably needs to be fixed to use a properly translated string, but it'd be OK to open a separate issue for it since it's not *directly* part of making emails translatable.

### PR DESCRIPTION
**Pull request by Leats**
Here comes a big one :blush:

Closes #5263.

### CODE REVIEW

File: indico/modules/events/static/notifications_test.py [hunk 1]
```diff
@@ -25,4 +25,5 @@ def test_paper_judgement_email_plaintext(snapshot, dummy_user, dummy_contributio
     template = get_template_module('events/static/emails/paper_judgment_notification_email.txt',
                                    paper=revision, contribution=dummy_contribution)
     snapshot.snapshot_dir = Path(__file__).parent / 'templates/emails/tests'
+    print(template.get_body())
```
Comment (pferreir) [inline indico/modules/events/static/notifications_test.py:28]: Forgotten `print`.

File: indico/modules/rb/models/blocked_rooms_test.py [hunk 2]
```diff
@@ -33,7 +33,8 @@ def test_reject(dummy_user, dummy_blocking, smtp, with_user, with_reason):
         kwargs['reason'] = 'foo'
     br.reject(**kwargs)
     assert br.state == BlockedRoom.State.rejected
-    mail = extract_emails(smtp, one=True, to=dummy_blocking.created_by_user.email, subject='Room blocking REJECTED')
+    mail = extract_emails(smtp, one=True, to=dummy_blocking.created_by_user.email,
+                          subject='[Indico] Room blocking REJECTED')
```
Comment (pferreir) [inline indico/modules/rb/models/blocked_rooms_test.py:37]: Isn't there a translation call missing here?

Comment (ThiefMaster) [reply indico/modules/rb/models/blocked_rooms_test.py:37]: unit test ;)

Comment (pferreir) [reply indico/modules/rb/models/blocked_rooms_test.py:37]: whoops. *hides*

File: indico/modules/rb/templates/emails/reservations/creation_email_to_user.txt [hunk 3]
```diff
@@ -1,31 +1,76 @@
...
+            [{{ room }}] Pre-Booking awaiting acceptance ({{ start }})
+        {%- endtrans %}
+    {%- else -%}
+        {% trans room=reservation.room.full_name, start=reservation.start_dt|format_datetime('dd/MM/yyyy HH:mm', locale='en_GB') -%}
+            [{{ room }}] Booking confirmed ({{ start }})
+        {%- endtrans %}
+    {%- endif %}
+{%- endblock %}
+
+{% block email_body -%}
```
Comment (ThiefMaster) [inline indico/modules/rb/templates/emails/reservations/creation_email_to_user.txt:17]: I think the whole block can be wrapped in a `dedent` filter in order to properly indent all the conditionals and make them readable ;)

File: indico/modules/rb/templates/emails/reservations/modification_email_to_manager.txt [hunk 4]
```diff
@@ -1,5 +1,9 @@
 {% extends 'rb/emails/reservations/base_email_to_manager.txt' %}
 
-{% block body -%}
-A booking in a room you manage has been MODIFIED.
+{% block subject -%}{% trans room=reservation.room.full_name, start=reservation.start_dt|format_datetime('dd/MM/yyyy HH:mm', locale='en_GB') -%}
+    [{{ room }}] Booking modified ({{ start }})
+{%- endtrans %}{%- endblock %}
+
+{% block email_body -%}
+{% trans %}A booking in a room you manage has been MODIFIED.{% endtrans %}
```
Comment (ThiefMaster) [inline indico/modules/rb/templates/emails/reservations/modification_email_to_manager.txt:8]: can be indented - the block is configured to swallow whitespace (`-%} ... {%-`)

File: indico/modules/rb/templates/emails/reservations/reservation_info.txt [hunk 5]
```diff
@@ -1,5 +1,5 @@
-Room:   {{ reservation.room.full_name }}
-For:    {{ reservation.booked_for_name }}
-Reason: {{ reservation.booking_reason }}
-Dates:  {{ reservation.start_dt | format_date('EEEE dd/MM/yyyy', locale='en_GB') }} - {{ reservation.end_dt | format_date('EEEE dd/MM/yyyy', locale='en_GB') }}
-Hours:  {{ reservation.start_dt | format_time(locale='en_GB') }} - {{ reservation.end_dt | format_time(locale='en_GB') }}
+{% trans %}Room:{% endtrans %}   {{ reservation.room.full_name }}
```
Comment (ThiefMaster) [inline indico/modules/rb/templates/emails/reservations/reservation_info.txt:1]: Probably another place where aligning values no longer works well :/

File: indico/modules/events/static/templates/emails/paper_judgment_notification_email.txt [hunk 6]
```diff
@@ -1,38 +1,51 @@
-{% extends 'emails/base.txt' %}
```
[the remark this answers is not in the document]
Comment (Leats) [reply indico/modules/events/static/templates/emails/paper_judgment_notification_email.txt:1]: To be fair I hadn't found a spot where that email got referenced anyway (but I had commented on that already in PMs)

File: indico/modules/rb/notifications/notifications_test.py [hunk 7]
```diff
@@ -85,15 +86,17 @@ def test_reservation_confirmation_emails_plaintext(snapshot, snapshot_name, crea
...
+    ('creation_email_to_user.txt', '_pre', None, RepeatFrequency.NEVER),
+    ('creation_email_to_user.txt', '_repeat_pre', None, RepeatFrequency.DAY),
 ))
 def test_reservation_creation_emails_plaintext(
     snapshot, snapshot_name, create_reservation, suffix, excluded_days, repeat
 ):
     res = create_reservation(id=1337, start_dt=datetime(2022, 11, 11, 13, 37), end_dt=datetime(2022, 11, 21, 14, 37),
                              repeat_frequency=repeat)
+    if 'pre' in suffix:
+        res.state = ReservationState.pending
```
[the remark this answers is not in the document]
Comment (Leats) [reply indico/modules/rb/notifications/notifications_test.py:99]: Yeah me neither tbh... so if you want me to change it I will, I just didn't like the alternatives as well :D

File: indico/modules/rb/notifications/reservations.py [hunk 8]
```diff
@@ -46,48 +46,38 @@ def __init__(self, reservation):
...
 
     def compose_email_to_user(self, **mail_params):
         creator = self.reservation.created_by_user
         to_list = {creator.email}
         if self.reservation.contact_email:
             to_list.add(self.reservation.contact_email)
-        subject = self._get_email_subject(**mail_params)
-        body = self._make_body(mail_params, reservation=self.reservation)
-        return make_email(to_list=to_list, subject=subject, body=body)
+        with force_user_locale(creator):
```
[the remark this answers is not in the document]
Comment (Leats) [reply indico/modules/rb/notifications/reservations.py:62]: Hm true. Tbh right now I'd keep it similar to managers and just force default locale? But obv that's switchable

File: indico/modules/auth/controllers.py [hunk 9]
```diff
@@ -137,8 +137,9 @@ def _send_confirmation(email, salt, endpoint, template, template_args=None, url_
     url_args = url_args or {}
     token = secure_serializer.dumps(data or email, salt=salt)
     url = url_for(endpoint, token=token, _external=True, **url_args)
-    template_module = get_template_module(template, email=email, url=url, **template_args)
-    send_email(make_email(email, template=template_module))
+    with force_locale(None) if 'user' not in template_args else template_args['user'].force_user_locale():
```
Comment (ThiefMaster) [inline indico/modules/auth/controllers.py:140]: yikes... (but i can't come up with a less ugly solution for this either ;x)

File: indico/modules/users/templates/emails/verify_email.txt [hunk 10]
```diff
@@ -5,13 +5,13 @@
...
 {% block header_recipient -%}
-    {{ user.first_name }}
+    {% trans name=user.first_name %}Dear {{ name }},{% endtrans %}
 {%- endblock %}
 
 {% block body -%}
-{%- trans notrimmed -%}
-You recently added the e-mail address {{ email }} to your Indico account.
-Please click the following link to confirm this action.
+{%- trans notrimmed mail=email -%}
```
Comment (ThiefMaster) [inline indico/modules/users/templates/emails/verify_email.txt:12]: Why is this needed? Using `{{ email }}` below w/o the assignment here should work as well..

Comment (ThiefMaster) [inline indico/modules/users/templates/emails/verify_email.txt:12]: I'd ditch the `notrimmed` everywhere (having significant whitespace in strings exposed to translators is a mess) and wrap it in a wordwrap filter block instead. (I thought I suggested this before, or maybe I just wanted to and forgot to save? 🙈)

File: indico/modules/users/models/users.py [hunk 11]
```diff
@@ -783,6 +784,13 @@ def get_merged_from_users_recursive(self):
         cte = cte_query.union_all(rec_query)
         return User.query.join(cte, User.id == cte.c.merged_from_id).all()
 
+    @contextmanager
+    def force_user_locale(self):
+        """Temporarily override the locale to the locale of the user."""
+        locale = self.settings.get('lang') or None
```
Comment (ThiefMaster) [inline indico/modules/users/models/users.py:790]: do we sometimes have a falsy value there that's not `None`? if not the `or None` isn't needed :)

File: indico/modules/rb/templates/emails/reservations/reminders/upcoming_occurrence.html [hunk 12]
```diff
@@ -30,14 +30,18 @@
...
-        If you do not need to use the room, please
-        <a href="{{ reservation.external_details_url }}">cancel the reservation in Indico</a>.
+        {% trans a=' <a href="{{ reservation.external_details_url }}">'|safe enda='</a>'|safe %}
+            If you do not need to use the room, please
+            {{ a }}cancel the reservation in Indico{{ enda }}.
+        {%- endtrans -%}
     {%- else -%}
-        If you do not need to use the room, please
-        <a href="{{ occurrence.external_cancellation_url }}">cancel the occurrence in Indico</a>.
+        {% trans a=' <a href="{{ occurrence.external_cancellation_url }}">'|safe enda='</a>'|safe %}
```
Comment (ThiefMaster) [inline indico/modules/rb/templates/emails/reservations/reminders/upcoming_occurrence.html:42]: I'd use link/endlink to make it less confusing for translators. Is the leading whitespace needed BTW? The linebreak should already cause a whitespace.

File: indico/modules/rb/templates/emails/reservations/creation_email_to_user.txt [hunk 13]
```diff
@@ -1,31 +1,78 @@
...
-{%- block confirmed_booking -%}
-{%- if excluded_days %}
-    (Note that there are {{ excluded_days }} excluded days. For further info, check your reservation)
-{%- endif %}
+{% block email_body -%}
+    {%- filter dedent -%}
+        {%- block booking_details -%}
+            {% if prebooking -%}
+                {% if reservation.repeat_frequency == RepeatFrequency.NEVER -%}
+                    {% trans notrimmed room=reservation.room.full_name, name=reservation.booked_for_name, date=reservation.start_dt | format_date('EEEE dd/MM/yyyy', locale='en_GB'), from=reservation.start_dt | format_time(locale='en_GB'), to=reservation.end_dt | format_time(locale='en_GB') -%}
```
Comment (ThiefMaster) [inline indico/modules/rb/templates/emails/reservations/creation_email_to_user.txt:22]: - Can we wrap this over multiple lines? 🙈 
- Removing the whitespace around the `|` filter operator makes this more readable (the grouping feels really weird when reading it right now)
- I don't think forcing fixed date formats and a locale is a good idea here (we don't want dd/mm/yyyy format in en_US for example, and french speakers probably don't want english weekday names ;)). Try `|format_skeleton('EEEEddMMyyyy')` instead, this is smart enough to pick the closest format for the current locale that matches the sleketon:

```python
In [11]: format_skeleton(dt(2022, 9, 11), 'EEEEddMMyyyy', 'en_GB', 'Europe/Zurich')
Out[11]: 'Sunday, 11/09/2022'

In [12]: format_skeleton(dt(2022, 9, 11), 'EEEEddMMyyyy', 'de_DE', 'Europe/Zurich')
Out[12]: 'Sonntag, 11.09.2022'

In [13]: format_skeleton(dt(2022, 9, 11), 'EEEEddMMyyyy', 'fr_FR', 'Europe/Zurich')
Out[13]: 'dimanche 11/09/2022'

In [14]: format_skeleton(dt(2022, 9, 11), 'EEEEddMMyyyy', 'en_US', 'Europe/Zurich')
Out[14]: 'Sunday, 09/11/2022'
```

File: indico/modules/events/registration/templates/emails/registration_state_update_to_managers.html [hunk 14]
```diff
@@ -2,12 +2,46 @@
 {% block registration_body %}{% endblock %}
 
 {% block subject_message -%}
-    Registration {{ registration.state.title|lower }}
+    {% if registration.state.name == 'complete' -%}
+        {% trans title=event.title %}Registration complete for {{ title }}{% endtrans%}
```
Comment (ThiefMaster) [inline indico/modules/events/registration/templates/emails/registration_state_update_to_managers.html:6]:
```suggestion
        {% trans title=event.title %}Registration complete for {{ title }}{% endtrans %}
```

same below...

File: indico/modules/events/payment/templates/emails/payment_inconsistency_email_to_manager.txt [hunk 15]
```diff
@@ -1,7 +1,19 @@
...
+    {% trans %}Payment inconsistency{% endtrans %}
+{%- endblock %}
 
-Regards.
+{% block header -%}
+    {% trans title=event.title %}Dear manager of the event {{ title }},{% endtrans %}
+{%- endblock %}
+
+{% block body %}
+{%filter dedent -%}
```
Comment (ThiefMaster) [inline indico/modules/events/payment/templates/emails/payment_inconsistency_email_to_manager.txt:12]:
```suggestion
{% filter dedent -%}
```

(maybe uses the regex searches i sent you again :p)

File: indico/modules/events/papers/templates/emails/paper_assignment.html [hunk 16]
```diff
@@ -2,34 +2,42 @@
...
 {% block content_header %}
-    <p>Dear {{ receiver.first_name }},</p>
+    <p>{% trans name=receiver.first_name %}Dear {{name }},{% endtrans %}</p>
     {% if contribs | length == 1 %}
         {% set contrib = contribs | first %}
         {% if assign %}
-            <p>You have been assigned as a <strong>{{ role.title }}</strong> to the paper <strong>{{ contrib.title }}</strong>.</p>
+            <p>{% trans strong='<strong>'|safe, endstrong='</strong>'|safe, title=contrib.title, role_title=role.title -%}
+                You have been assigned as a {{ strong }}{{ role_title }}{{ endstrong }} to the paper {{ strong }}{{ title }}{{ endstrong }}.
+            {%- endtrans %}</p>
```
Comment (ThiefMaster) [inline indico/modules/events/papers/templates/emails/paper_assignment.html:26]:
```suggestion
            <p>
                {%- trans strong='<strong>'|safe, endstrong='</strong>'|safe, title=contrib.title, role_title=role.title -%}
                    You have been assigned as a {{ strong }}{{ role_title }}{{ endstrong }} to the paper {{ strong }}{{ title }}{{ endstrong }}.
                {%- endtrans -%}
            </p>
```

File: indico/modules/events/papers/notifications_test.py [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+# modify it under the terms of the MIT License; see the
+# LICENSE file for more details.
+
+from pathlib import Path
+
+from indico.modules.events.papers.models.revisions import PaperRevisionState
+from indico.web.flask.templating import get_template_module
+
+
+def test_paper_judgement_email_plaintext(snapshot, dummy_user, dummy_contribution):
```
Comment (ThiefMaster) [inline indico/modules/events/papers/notifications_test.py:14]:
```suggestion
def test_paper_judgment_email_plaintext(snapshot, dummy_user, dummy_contribution):
```

File: indico/modules/events/registration/templates/emails/registration_state_update_to_managers.html (hunk 14 above)
Comment (Leats) [reply indico/modules/events/registration/templates/emails/registration_state_update_to_managers.html:6]: wtf i was so sure I had replaces all of them...

File: indico/modules/events/payment/templates/emails/payment_inconsistency_email_to_manager.txt (hunk 15 above)
Comment (Leats) [reply indico/modules/events/payment/templates/emails/payment_inconsistency_email_to_manager.txt:12]: I DID.... did I forget to push at some point?? wtf...

File: indico/modules/rb/notifications/blockings.py [hunk 18]
```diff
@@ -28,9 +27,9 @@ def notify_request_response(blocked_room):
...
-    to = blocked_room.blocking.created_by_user.email
+    user = blocked_room.blocking.created_by_user
     verb = blocked_room.State(blocked_room.state).title.upper()
-    subject = f'Room blocking {verb}'
-    body = render_template('rb/emails/blockings/state_email_to_user.txt',
-                           blocking=blocked_room.blocking, blocked_room=blocked_room, verb=verb)
-    return make_email(to, subject=subject, body=body)
+    with user.force_user_locale():
+        tpl = get_template_module('rb/emails/blockings/state_email_to_user.txt',
+                                  blocking=blocked_room.blocking, blocked_room=blocked_room, verb=verb)
```
Comment (ThiefMaster) [inline indico/modules/rb/notifications/blockings.py:34]: The `verb` can go away altogether now since you are no longer using it

File: indico/modules/rb/notifications/notifications_test.py [hunk 19]
```diff
@@ -46,7 +47,7 @@ def test_blocking_confirmation_email_plaintext(snapshot, dummy_room, dummy_user,
 def test_blocking_state_email_plaintext(snapshot, create_blocking, dummy_room, suffix, state):
     blocking = create_blocking(id=2)
     _assert_snapshot(snapshot, 'state_email_to_user.txt', 'blockings', suffix,
-                     blocking=blocking, verb='forgotten',
+                     blocking=blocking, verb=state.title.upper(),
```
Comment (ThiefMaster) [inline indico/modules/rb/notifications/notifications_test.py:50]: probably no longer needed here either

File: indico/modules/events/surveys/templates/emails/new_submission_email.txt [hunk 20]
```diff
@@ -1,27 +1,35 @@
-{% extends 'emails/base.txt' %}
+{% extends 'emails/base_i18n.txt' %}
 {% set survey = submission.survey %}
 
 {% block subject -%}
-    New survey submission for "{{ survey.title }}" on the event "{{ survey.event.title }}"
+    {% trans sur=survey.title, evt=survey.event.title %}New survey submission for "{{ sur }}" on the event "{{ evt }}"{% endtrans %}
```
Comment (ThiefMaster) [inline indico/modules/events/surveys/templates/emails/new_submission_email.txt:5]: I'd use meaningful names like `survey_title` and `event_title` - easier for translators if they know what a placeholder is used for. Same in the other survey templates which currently use the same abbreviations
